### PR TITLE
Unshare fixes

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Share ownership problems causes the issue linked below.

## Related Issue
https://github.com/owncloud/core/issues/26357

## How Has This Been Tested?
Manually checked that the PR fixes the linked issue. More tests incoming.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by jvillafanez** [comment]
* [ ] Moving the share to an external storage without allowing sharing behaves bad:

    ```
    curl -X MOVE -H "Destination: /remote.php/dav/files/qa2/Storage/007" -u qa2:11111 http://10.0.2.8:30080/remote.php/dav/files/qa2/base/sub1
    <?xml version="1.0" encoding="utf-8"?>
    <d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
      <s:exception>OCP\Share\Exceptions\GenericShareException</s:exception>
      <s:message>You are not allowed to share /qa2/files/Storage/007</s:message>
    </d:error>
    ```
    The folder (in this case) is moved to the external storage, but the share won't appear in the target user's folder (qa3 in this case). The source folder is still marked as shared

screenshot from 2017-11-22 17 43 22

* [x] Moving the share to an external storage allowing sharing in it works fine
* [ ] Moving the share to another share seems to work, but might need some explanation.

    Visually, the target share's owner (qa3 in this case) seems to be the owner of the files, but the DB still have the old share entry with the previous owner as share's owner.

To be more precise:

First test scenario:
1. Create 3 users: qa1, qa2 and qa3
2. Login with qa1, create a folder "base/sub" (first the parent folder and then the subfolder) and upload some files there.
3. Share "base" folder with qa2
4. Login as qa2
5. As qa2 setup a personal external storage without sharing permissions
6. As qa2 go to the "base" folder (shared by qa1) and share "sub" with qa3.
7. Login as qa3 and make sure qa3 has access to the contents of the "base/sub" folder
8. As qa2 move the "base/sub" folder (which is shared with qa3) to the personal external storage.

Third test scenario:
1. Create 3 users: qa1, qa2 and qa3
2. Login with qa1, create a folder "base/sub" (first the parent folder and then the subfolder) and upload some files there.
3. Share "base" folder with qa2
4. Login as qa2
5. As qa2 go to the base folder and share "sub" with qa3
6. Login as qa3 and verify the "sub" folder is visible
7. As qa3 create a "qa3owner" folder and share it with qa2
8. As qa2 move the "base/sub" folder to the "qa3owner" folder which should be visible for qa2

File: apps/dav/lib/DAV/FileCustomPropertiesBackend.php [hunk 1]
```diff
@@ -49,12 +49,23 @@ class FileCustomPropertiesBackend extends AbstractCustomPropertiesBackend {
 	const DELETE_BY_ID_AND_NAME_STMT = 'DELETE FROM `*PREFIX*properties`'
 	. ' WHERE `fileid` = ? AND `propertyname` = ?';
 
+	private $moveSource = null;
```
Comment (PVince81) [inline apps/dav/lib/DAV/FileCustomPropertiesBackend.php]: PHPDoc, what is this ? what type ?

**Comment by PVince81** [comment]
> Moving the share to an external storage without allowing sharing behaves bad:

Likely a separate bug. I'm not sure what the correct behavior should be. Maybe raise a separate ticket with steps to reproduce and we can discuss expected behavior there.

**Comment by PVince81** [comment]
> As qa3 create a "qa3owner" folder and share it with qa2

Is "qa3owner" created in the root of qa3 ?

For the third scenario we need to change the share owner in the database.
I think this already happens for simpler sharing scenarios so the code might exist somewhere.

**Comment by PVince81** [comment]
Code looks good so far

**Comment by jvillafanez** [comment]
I think this is ready for review. As said, I'll open a new ticket for the first scenario. The third scenario should be fixed with the last commit.

There is another ticket to discuss regarding the third scenario.

1. Create 3 users: qa1, qa2 and qa3
2. Login with qa1, create a folder "base/sub" (first the parent folder and then the subfolder) and upload some files there.
3. Share "base" folder with qa2
4. Login as qa2
5. As qa2 go to the base folder and share "sub" with qa3 **and admin** (<--- this is new)
6. Login as qa3 and verify the "sub" folder is visible
7. As qa3 create a "qa3owner" folder and share it with qa2
8. As qa2 move the "base/sub" folder to the "qa3owner" folder which should be visible for qa2

The problem is that qa3 doesn't know that the "qa3owner/sub" folder is shared with another user. There is no visual indicator of that unless the qa3 user clicks on the folder. The same happens with simpler scenarios where a shared file is reshared by another user to a third one.

**Comment by jvillafanez** [comment]
Unittests incoming

**Comment by PVince81** [comment]
> The problem is that qa3 doesn't know that the "qa3owner/sub" folder is shared with another user. There is no visual indicator of that unless the qa3 user clicks on the folder. The same happens with simpler scenarios where a shared file is reshared by another user to a third one.

Can you make yet another ticket for this one ? I observed this behavior before a few days ago. I hope it's only an indicator thing and that the DB structure looks fine already.

File: apps/files_sharing/lib/Updater.php [hunk 2]
```diff
@@ -68,16 +68,22 @@ static private function moveShareToShare($path) {
...
 		}
 
-		$newOwner = $dstMount->getShare()->getShareOwner();
 
 		//Ownership is moved over
 		foreach ($shares as $share) {
 			/** @var \OCP\Share\IShare $share */
-			$share->setShareOwner($newOwner);
-			$shareManager->updateShare($share);
+			if ($share->getShareWith() !== $newOwner) {
```
Comment (PVince81) [inline apps/files_sharing/lib/Updater.php]: Maybe at some point we'll want a `$shareManager->changeOwner()` with all the above and below logic embedded. At some point in the future we might have such action triggerable over the UI.

This is fine for now.

**Comment by jvillafanez** [comment]
Looks like there won't be new unittests... I'm not sure what is the setup for the UpdaterTest.php tests (files_sharing app) but I'm having a lot of problems trying to move around a shared file from another account.
The rename operation fails with:
```
{"reqId":"eDRNRo3ESbFAY5Nc4PVD","level":3,"time":"2017-11-28T12:16:16+00:00","re
moteAddr":"","user":"test-share-user2","app":"core","method":"--","url":"--","me
ssage":"unable to rename, source directory is not writable : files\/folder_share
_updater_test\/base1"}
```
I've removed the cleanup operations from the tests to check what is the state after the test: DB users are missing, files and shares for the accounts are there. After creating all the required users, the FS state is the expected one, and I can move the shared file around without problems.

I guess it might be better to test these scenarios via integration tests.

**Comment by PVince81** [comment]
> I guess it might be better to test these scenarios via integration tests.

Sounds good, thanks

**Comment by jvillafanez** [comment]
This should be ready to be merged.

Backport to stable10?

**Review by PVince81 — APPROVED** [review]
Awesome, thanks :+1:

Please backport to stable10 for 10.0.5

**Comment by jvillafanez** [comment]
Backport in https://github.com/owncloud/core/pull/29716

**Comment by jvillafanez** [comment]
Tickets opened:
https://github.com/owncloud/core/issues/29717
https://github.com/owncloud/core/issues/29718

**Comment by PVince81** [comment]
@jvillafanez codecov is not happy.

Adding tests for the DAV properties thing should be easy.

> Looks like there won't be new unittests... I'm not sure what is the setup for the UpdaterTest.php tests 

Maybe setup another test class that only works with mocks instead of the integration nightmare of the old tests ?

**Comment by jvillafanez** [comment]
> Adding tests for the DAV properties thing should be easy.

I guess not for me... Other than adding the tests just for the coverage, whose worthyness will be under question, I'm not sure how to test the scenario.
```
$fileCustomPropertyBackend->move(....);
$fileCustomPropertyBackend->delete(....);
// checks
```
They'll be "I'm not sure what I'm testing" kind of tests. It makes me wonder if the fix for that problem should be in another place.

> Maybe setup another test class that only works with mocks instead of the integration nightmare of the old tests ?

It's a static class with heavy server component dependencies. Even if the we manage to mock all the required server components, we'll likely need to mock more subcomponents.

**Comment by PVince81** [comment]
Ok, fine

**Pull request merged**